pystac/utils.py

```diff
@@ -31,18 +37,89 @@ def _urlparse(href: str) -> URLParseResult:
         return parsed
 
 
-def _join(is_path: bool, *args: str) -> str:
+class JoinType(str, Enum):
+    """Allowed join types for the :func:`_join` function."""
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    @staticmethod
+    def from_parsed_uri(parsed_uri: URLParseResult) -> "JoinType":
+        """Determines the appropriate join type based on the scheme of the parsed
+        result."""
+        if parsed_uri.scheme == "":
+            return JoinType.PATH
+        else:
+            return JoinType.URL
+
+    PATH = "path"
+    URL = "url"
+
+
+def join_path_or_url(join_type: JoinType, *args: str) -> str:
     """Version of os.path.join that takes into account whether or not we are working
-    with a URL.
+    with a URL."""
 
-    A windows system shouldn't use os.path.join if we're working with a URL.
-    """
-    if is_path:
+    if join_type == JoinType.PATH:
         return _pathlib.join(*args)
     else:
         return posixpath.join(*args)
 
 
+def _make_relative_href_url(
+    parsed_source: URLParseResult,
+    parsed_start: URLParseResult,
+    start_is_dir: bool = False,
+) -> str:
+
+    # If the start path is not a directory, get the parent directory
+    start_dir = (
+        parsed_start.path if start_is_dir else parsed_start.path.rsplit("/", 1)[0]
+    )
+
+    # Strip the leading slashes from both paths
+    start_dir = start_dir.lstrip("/")
+    source_path = parsed_source.path.lstrip("/")
+
+    # Get the relative path
+    rel_url = posixpath.relpath(source_path, start_dir)
+
+    # Ensure we retain a trailing slash from the original source path
+    if parsed_source.path.endswith("/"):
+        rel_url += "/"
+
+    # Prepend the "./"
+    if not rel_url.startswith("."):
```

Comment:
I believe this would break if any of the filenames start with a dot:

```
>>> posixpath.relpath("/foo/.bar", "/foo")
'.bar'
```

Could you do `if rel_url != "./" and not rel_url.startswith("../"):`? I think that should handle the two relevant cases, by making sure the string match includes a slash (which can't be part of a filename).

Reply:
Thanks for catching this. I'll add a test case for this as well.

Reply:
Tested this case in https://github.com/duckontheweb/pystac/pull/1, which does not fail. Perhaps because of the above logic on the parsed startdir this isn't an issue?

Reply:
Added a test case and updated the logic in cff9606a056ab9107a71da9cc7128a5b13a8fcd1